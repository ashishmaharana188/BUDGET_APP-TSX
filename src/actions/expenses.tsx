import { v4 as uuid } from "uuid";

// ADD_EXPENSE
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdDt = 0,
} = {}) => ({
  type: "ADD_EXPENSE" as const,
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdDt,
  },
});
// REMOVE_EXPENSE
export const removeExpense = (id: string) => ({
  type: "REMOVE_EXPENSE" as const,
  id,
});
// EDIT_EXPENSE
export const editExpense = (id: string, updates: any) => ({
  type: "EDIT_EXPENSE" as const,
  id,
  updates,
});
