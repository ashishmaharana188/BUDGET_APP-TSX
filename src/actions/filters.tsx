import moment from "moment";

// SET_TEXT_FILTER
export const setTextFilter = (text: string = "") => ({
  type: "SET_TEXT_FIELD",
  text,
});
export const setAmountFilter = (amount: string = "") => ({
  type: "SET_AMOUNT_FIELD",
  amount,
});
// SORT_BY_AMOUNT
export const sortByAmount = (sortOrder: "asc" | "desc" | null) => ({
  type: "SORT_BY_AMOUNT",
  payload: { sortOrder },
});
export const setSortBy = (sortBy: "date" | "amount" | null) => ({
  type: "SET_SORT_BY",
  sortBy,
});
// SORT_BY_DATE
export const sortByDate = (sortOrder: "asc" | "desc" | null) => ({
  type: "SORT_BY_DATE",
  payload: { sortOrder },
});
// SET_START_DATE
export const setStartDate = (startDate?: number | null) => ({
  type: "SET_START_DATE",
  startDate:
    startDate !== null && startDate !== undefined
      ? moment(startDate).valueOf()
      : null,
});
// SET_END_DATE
export const setEndDate = (endDate?: number | null) => ({
  type: "SET_END_DATE",
  endDate:
    endDate !== null && endDate !== undefined
      ? moment(endDate).valueOf()
      : null,
});
export const resetFilters = () => ({
  type: "RESET_FILTERS",
});
