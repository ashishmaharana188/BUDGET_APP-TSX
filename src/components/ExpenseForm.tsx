import React from "react";
import moment, { Moment } from "moment";
import { DatePicker, LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterMoment } from "@mui/x-date-pickers/AdapterMoment";
import {
  ExpenseFormProps,
  ExpenseFormState,
} from "../TS_Interface/GlobalInterface";

export default class ExpenseForm extends React.Component<
  ExpenseFormProps,
  ExpenseFormState
> {
  constructor(props: any) {
    super(props);
    this.state = {
      description: props.expense?.description || "",
      note: props.expense?.note || "",
      amount: props.expense ? props.expense.amount.toString() : "",
      createdDt: props.expense ? moment(props.expense.createdDt) : moment(),
      error: "",
    };
  }

  onDescriptionChange = (e: any) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e: any) => {
    const note = e.target.value.replace(/\n/g, " ");
    this.setState(() => ({ note }));
  };
  onAmountChange = (e: any) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onAmountBlur = () => {
    const { amount } = this.state;
    const parsedAmount = parseFloat(amount);
    // Check if the parsed amount is a valid number
    if (!isNaN(parsedAmount)) {
      // Check if the amount is a whole number
      const formattedValue = Number.isInteger(parsedAmount)
        ? parsedAmount.toFixed(2) // Format whole number to two decimal points
        : amount; // Keep the decimal number as is
      this.setState(() => ({ amount: formattedValue }));
    }
  };

  onDateChange = (createdDt: Moment | null) => {
    if (createdDt) {
      this.setState(() => ({ createdDt }));
    }
  };
  onFocus = () => {
    // Set the date to today when the date picker is focused
    if (!this.state.createdDt) {
      this.setState(() => ({ createdDt: moment() }));
    }
  };
  onSubmit = (e: any) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide Description and Amount.",
      }));
      console.log();
    } else {
      this.setState(() => ({
        error: "",
      }));
      if (!this.state.createdDt) {
        this.setState(() => ({
          createdDt: moment(),
        }));
      }
      if (this.props.onSubmit) {
        this.props.onSubmit({
          description: this.state.description,
          amount: parseFloat(this.state.amount),
          createdDt: this.state.createdDt.valueOf(),
          note: this.state.note,
        });
      }
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description*"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount*"
            value={this.state.amount}
            onChange={this.onAmountChange}
            onBlur={this.onAmountBlur}
          />
          <LocalizationProvider dateAdapter={AdapterMoment}>
            <DatePicker
              value={this.state.createdDt}
              onChange={this.onDateChange}
            />
          </LocalizationProvider>
          <textarea
            placeholder="Add note"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
