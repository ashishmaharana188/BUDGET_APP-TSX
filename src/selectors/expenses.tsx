import { expenseReducerIntf } from "../TS_Interface/GlobalInterface";
import { filterReducerIntf } from "../TS_Interface/GlobalInterface";
import moment from "moment";

export default (
  expenses: expenseReducerIntf[],
  { text, amount, sortBy, startDate, endDate, sortOrder }: filterReducerIntf
) => {
  return expenses
    .filter((expense) => {
      const createdDtMoment = moment(expense.createdDt);

      const startDateMatch = startDate
        ? moment(startDate).isSameOrBefore(createdDtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? moment(endDate).isSameOrAfter(createdDtMoment, "day")
        : true;
      /* const textMatch = text
        ? expense.description.toLowerCase().includes(text.toLowerCase())
        : true;

      const amountMatch = !isNaN(parseFloat(amount))
        ? expense.amount === parseFloat(amount)
        : true;  

      return startDateMatch && endDateMatch && (textMatch || amountMatch);*/
      const isNumber = !isNaN(parseFloat(text)) && isFinite(Number(text));
      if (isNumber) {
        return (
          startDateMatch && endDateMatch && expense.amount === parseFloat(text)
        );
      } else {
        return (
          startDateMatch &&
          endDateMatch &&
          expense.description.toLowerCase().includes(text.toLowerCase())
        );
      }
    })
    .sort((a, b) => {
      if (sortBy === "date" && sortOrder === "desc") {
        return a.createdDt < b.createdDt ? 1 : -1; // Descending by date
      } else if (sortBy === "date" && sortOrder === "asc") {
        return a.createdDt > b.createdDt ? 1 : -1; // Ascending by date
      } else if (sortBy === "amount" && sortOrder === "asc") {
        return a.amount - b.amount; // Ascending by amount
      } else if (sortBy === "amount" && sortOrder === "desc") {
        return b.amount - a.amount; // Descending by amount
      } else {
        return 0; // No sorting
      }
    });
};
