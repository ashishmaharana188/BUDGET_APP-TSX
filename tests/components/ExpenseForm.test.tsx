import React from "react";
import { test, expect } from "vitest";
import {
  render,
  fireEvent,
  getByLabelText,
  getByPlaceholderText,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import ExpenseForm from "../../src/components/ExpenseForm";
import { addExpense } from "./../../src/actions/expenses";
import expenses from "../../tests/fixtures/expense";

test("should render invalid form", async () => {
  const { getByPlaceholderText, getByText, asFragment } = render(
    <MemoryRouter>
      <ExpenseForm />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("should render one Expense correctly", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <ExpenseForm expense={expenses[0]} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("should render error with expense data", async () => {
  const { asFragment, getByPlaceholderText, getByText } = render(
    <MemoryRouter>
      <ExpenseForm />
    </MemoryRouter>
  );
  const descriptionField = getByPlaceholderText(/Description/i);
  const amountField = getByPlaceholderText(/Amount/i); // Replace with your actual input label

  const submitButton = getByText(/Add Expense/i);

  await userEvent.click(submitButton);

  const errorMessage = getByText(/Please provide Description and Amount/i);
  expect(errorMessage).to.exist;

  expect(asFragment()).toMatchSnapshot();
});

test("should render description properly with error for amount", async () => {
  const { getByPlaceholderText, getByText, asFragment } = render(
    <MemoryRouter>
      <ExpenseForm />
    </MemoryRouter>
  );

  const descriptionField = getByPlaceholderText(/Description/i);
  const amountField = getByPlaceholderText(/Amount/i); // Replace with your actual input label

  await userEvent.type(descriptionField, "Testing_1");

  const submitButton = getByText(/Add Expense/i);
  await userEvent.click(submitButton);

  const errorMessage = getByText(/Please provide an Amount./i);
  expect(errorMessage).to.exist;

  expect(asFragment()).toMatchSnapshot();
});

test("should render description properly with no error", async () => {
  const { getByPlaceholderText, getByText, asFragment } = render(
    <MemoryRouter>
      <ExpenseForm />
    </MemoryRouter>
  );
  const descriptionField = getByPlaceholderText(/Description/i);
  const amountField = getByPlaceholderText(/Amount/i); // Replace with your actual input label

  await userEvent.type(descriptionField, "Testing_1");
  await userEvent.type(amountField, "1000");

  const submitButton = getByText(/Add Expense/i);
  await userEvent.click(submitButton);

  expect(asFragment()).toMatchSnapshot();
});

test("should render notes properly with error", async () => {
  const { getByPlaceholderText, getByText, asFragment } = render(
    <MemoryRouter>
      <ExpenseForm />
    </MemoryRouter>
  );
  const noteField = getByPlaceholderText(/Note/i);

  await userEvent.type(noteField, "Testing_note");

  const submitButton = getByText(/Add Expense/i);
  await userEvent.click(submitButton);

  const errorMessage = getByText(/Please provide Description and Amount/i);
  expect(errorMessage).to.exist;

  expect(asFragment()).toMatchSnapshot();
});
