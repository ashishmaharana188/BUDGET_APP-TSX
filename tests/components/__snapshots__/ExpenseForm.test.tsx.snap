// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render description properly with error for amount 1`] = `
<DocumentFragment>
  <div>
    <p>
      Please provide Description and Amount.
    </p>
    <form>
      <input
        placeholder="Description*"
        type="text"
        value="Testing_1"
      />
      <input
        placeholder="Amount*"
        type="text"
        value=""
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1b1fjlj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r6:"
            inputmode="text"
            placeholder="MM/DD/YYYY"
            type="text"
            value="01/14/2025"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 14, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <textarea
        placeholder="Add note"
      />
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should render description properly with no error 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        placeholder="Description*"
        type="text"
        value="Testing_1"
      />
      <input
        placeholder="Amount*"
        type="text"
        value="1000.00"
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1b1fjlj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r8:"
            inputmode="text"
            placeholder="MM/DD/YYYY"
            type="text"
            value="01/14/2025"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 14, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <textarea
        placeholder="Add note"
      />
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should render error with expense data 1`] = `
<DocumentFragment>
  <div>
    <p>
      Please provide Description and Amount.
    </p>
    <form>
      <input
        placeholder="Description*"
        type="text"
        value=""
      />
      <input
        placeholder="Amount*"
        type="text"
        value=""
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1b1fjlj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r4:"
            inputmode="text"
            placeholder="MM/DD/YYYY"
            type="text"
            value="01/14/2025"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 14, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <textarea
        placeholder="Add note"
      />
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should render invalid form 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        placeholder="Description*"
        type="text"
        value=""
      />
      <input
        placeholder="Amount*"
        type="text"
        value=""
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1b1fjlj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            inputmode="text"
            placeholder="MM/DD/YYYY"
            type="text"
            value="01/14/2025"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 14, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <textarea
        placeholder="Add note"
      />
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should render notes properly with error 1`] = `
<DocumentFragment>
  <div>
    <p>
      Please provide Description and Amount.
    </p>
    <form>
      <input
        placeholder="Description*"
        type="text"
        value=""
      />
      <input
        placeholder="Amount*"
        type="text"
        value=""
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1b1fjlj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":ra:"
            inputmode="text"
            placeholder="MM/DD/YYYY"
            type="text"
            value="01/14/2025"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 14, 2025"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <textarea
        placeholder="Add note"
      >
        Testing_note
      </textarea>
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`should render one Expense correctly 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        placeholder="Description*"
        type="text"
        value="Weed"
      />
      <input
        placeholder="Amount*"
        type="text"
        value="100"
      />
      <div
        class="MuiFormControl-root MuiTextField-root css-1b1fjlj-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-jupps9-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            inputmode="text"
            placeholder="MM/DD/YYYY"
            type="text"
            value="01/01/1970"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date, selected date is Jan 1, 1970"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-oeq6yz-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <textarea
        placeholder="Add note"
      >
        high
      </textarea>
      <button>
        Add Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;
